include/alp/omp/blas3.hpp

```diff
@@ -0,0 +1,526 @@
+					std::is_same< InputType1, void >::value ||
+					std::is_same< InputType2, void >::value
+				),
+				"alp::internal::mxm_generic: the operator-monoid version of mxm cannot be "
+				"used if either of the input matrices is a pattern matrix (of type "
+				"void)"
+			);
+
+#ifndef NDEBUG
+			std::cout << "In alp::internal::mxm_generic (omp)\n";
+#endif
+
+			// Early exit checks 
+			if( ! internal::getInitialized( A ) ||
+				! internal::getInitialized( B ) ||
+				! internal::getInitialized( C )
+			) {
+				internal::setInitialized( C, false );
+				return SUCCESS;
+			}
+
+			const std::ptrdiff_t m   { static_cast< std::ptrdiff_t >( nrows( C ) ) };
+			const std::ptrdiff_t n   { static_cast< std::ptrdiff_t >( ncols( C ) ) };
+			const std::ptrdiff_t m_a { static_cast< std::ptrdiff_t >( nrows( A ) ) };
+			const std::ptrdiff_t k   { static_cast< std::ptrdiff_t >( ncols( A ) ) };
+			const std::ptrdiff_t k_b { static_cast< std::ptrdiff_t >( nrows( B ) ) };
+			const std::ptrdiff_t n_b { static_cast< std::ptrdiff_t >( ncols( B ) ) };
+
+			if( m != m_a || k != k_b || n != n_b ) {
+				return MISMATCH;
+			}
+
+			const Distribution_2_5D &da = internal::getAmf( A ).getDistribution();
+			const Distribution_2_5D &db = internal::getAmf( B ).getDistribution();
+			const Distribution_2_5D &dc = internal::getAmf( C ).getDistribution();
+
+			const auto tg_a = da.getThreadGridDims();
+			const auto tg_b = db.getThreadGridDims();
+			const auto tg_c = dc.getThreadGridDims();
+			
+			// 2.5D factor ranges within 3D limits
+			if( tg_c.rt != tg_a.rt || tg_a.rt != tg_b.rt || ( tg_a.tc % tg_a.rt ) != 0 ) {
+				return MISMATCH;
+			}
+
+			if( tg_c.tr != tg_a.tr || tg_c.tc != tg_b.tc || tg_a.tc != tg_b.tr ) {
+				return MISMATCH;
+			}
+
+			using th_coord_t = typename Distribution_2_5D::ThreadCoords;
+
+			RC rc = SUCCESS;
+
+#ifndef NDEBUG
+			#pragma omp critical
+			std::cerr << "In alp::internal::mxm_generic (omp)\n"
+				"\tBeginning parallel region" << std::endl;
+#endif
+
+			#pragma omp parallel
+			{			
+				const size_t thread = config::OMP::current_thread_ID();
+
+				const th_coord_t th_ijk_a = da.getThreadCoords( thread );
+				const th_coord_t th_ijk_b = db.getThreadCoords( thread );
+				const th_coord_t th_ijk_c = dc.getThreadCoords( thread );
+
+				RC local_rc = SUCCESS;
+
+// 				if( block_grid_dims_c.first != set_block_grid_dims_a.first 
+// 					|| block_grid_dims_c.second != set_block_grid_dims_b.second 
+// 					|| set_block_grid_dims_a.second != set_block_grid_dims_b.first 
+// 				) {
+// #ifndef NDEBUG
+// 					#pragma omp critical
+// 					std::cerr << "Thread " << thread << " in alp::internal::mxm_generic (omp)\n"
+// 						"\tMismatching local block grid size on set." << std::endl;
+// #endif
+// 					local_rc = MISMATCH;
+// 				}
+
+				// Broadcast A and B to all c-dimensional layers
+				if( local_rc == SUCCESS && da.isActiveThread( th_ijk_a ) && th_ijk_a.rt > 0 ) {
+
+#ifndef NDEBUG
+					#pragma omp critical
```

Comment:
Would it be better if each thread writes to its own files, rather than having `#pragma omp critical`?

Reply:
thread-specific file could be a good option actually. Still not sure what you mean with potential race-condition using omp critical? I'll note this down as a general comment for 558 in a separate issue so that maybe we can adopt it as a general style on this branch after the fix. Maybe also good to check currently estabished grb debugging style with shared mem perhaps we could uniform.